File: src/main/java/io/alexjoest/stackupup/core/ClassHierarchyRepository.java
package io.alexjoest.stackupup.core;

import java.util.concurrent.ConcurrentHashMap;

import net.minecraftforge.fml.common.asm.transformers.deobf.FMLDeobfuscatingRemapper;
import org.objectweb.asm.ClassReader;
import org.objectweb.asm.ClassVisitor;
import org.objectweb.asm.Opcodes;

import java.io.InputStream;
import java.util.Collections;
import java.util.LinkedHashSet;
import java.util.Set;

final class ClassHierarchyRepository {

    private ClassHierarchyRepository() {}

    private static final class Metadata {
        final String superClass;
        final Set<String> interfaces;

        Metadata(String superClass, Set<String> interfaces) {
            this.superClass = superClass;
            this.interfaces = interfaces;
        }
    }

    private static final ConcurrentHashMap<String, Metadata> cache = new ConcurrentHashMap<>();

    static String superClassOf(String className) {
        return get(className).superClass;
    }

    static Set<String> interfacesOf(String className) {
        return get(className).interfaces;
    }

    private static Metadata get(String className) {
        Metadata existing = cache.get(className);
        if (existing != null) {
            return existing;
        }
        Metadata loaded = loadMetadata(className);
        Metadata previous = cache.putIfAbsent(className, loaded);
        return previous != null ? previous : loaded;
    }

    private static Metadata loadMetadata(String className) {
        String fileName = FMLDeobfuscatingRemapper.INSTANCE.unmap(NameConverter.toSlashName(className));
        fileName = NameConverter.toSlashName(fileName) + ".class";
        InputStream stream = ClassHierarchyRepository.class.getClassLoader().getResourceAsStream(fileName);
        if (stream == null) {
            return new Metadata(null, Collections.<String>emptySet());
        }

        try {
            HierarchySignatureCollector collector = new HierarchySignatureCollector();
            ClassReader reader = new ClassReader(stream);
            reader.accept(collector, ClassReader.SKIP_CODE | ClassReader.SKIP_DEBUG | ClassReader.SKIP_FRAMES);
            return collector.toMetadata();
        } catch (Exception e) {
            return new Metadata(null, Collections.<String>emptySet());
        } finally {
            try { stream.close(); } catch (Exception ignored) {}
        }
    }

    private static class HierarchySignatureCollector extends ClassVisitor {
        private String rawSuperClass;
        private String[] rawInterfaces = new String[0];

        HierarchySignatureCollector() {
            super(Opcodes.ASM5);
        }

        @Override
        public void visit(int version, int access, String name, String signature,
                          String superName, String[] interfaces) {
            rawSuperClass = superName;
            rawInterfaces = interfaces;
        }

        Metadata toMetadata() {
            String superClass = null;
            if (rawSuperClass != null) {
                superClass = NameConverter.toDotName(FMLDeobfuscatingRemapper.INSTANCE.map(rawSuperClass));
            }
            LinkedHashSet<String> interfaces = new LinkedHashSet<>(rawInterfaces.length);
            for (String rawInterface : rawInterfaces) {
                interfaces.add(NameConverter.toDotName(FMLDeobfuscatingRemapper.INSTANCE.map(rawInterface)));
            }
            return new Metadata(superClass, interfaces);
        }
    }
}

File: src/main/kotlin/io/alexjoest/stackupup/core/ClassHierarchyRepository.kt
package io.alexjoest.stackupup.core

import net.minecraftforge.fml.common.asm.transformers.deobf.FMLDeobfuscatingRemapper
import org.objectweb.asm.ClassReader
import org.objectweb.asm.ClassVisitor
import org.objectweb.asm.Opcodes
import java.io.InputStream
import java.util.Collections
import java.util.LinkedHashSet
import java.util.concurrent.ConcurrentHashMap

private fun toSlashName(name: String): String {
    val builder = StringBuilder(name.length)
    for (char in name) {
        builder.append(if (char == '.') '/' else char)
    }
    return builder.toString()
}

private fun toDotName(name: String): String {
    val builder = StringBuilder(name.length)
    for (char in name) {
        builder.append(if (char == '/') '.' else char)
    }
    return builder.toString()
}

internal object ClassHierarchyRepository {
    private data class Metadata(
        val superClass: String?,
        val interfaces: Set<String>
    )

    private val cache = ConcurrentHashMap<String, Metadata>()

    fun superClassOf(className: String): String? = get(className).superClass

    fun interfacesOf(className: String): Set<String> = get(className).interfaces

    private fun get(className: String): Metadata {
        val existing = cache[className]
        if (existing != null) {
            return existing
        }
        val loaded = loadMetadata(className)
        val previous = cache.putIfAbsent(className, loaded)
        return previous ?: loaded
    }

    private fun loadMetadata(className: String): Metadata {
        var fileName = FMLDeobfuscatingRemapper.INSTANCE.unmap(toSlashName(className))
        fileName = toSlashName(fileName) + ".class"
        val stream: InputStream = ClassHierarchyRepository::class.java.classLoader.getResourceAsStream(fileName)
            ?: return Metadata(superClass = null, interfaces = Collections.emptySet())

        try {
            val collector = HierarchySignatureCollector()
            ClassReader(stream).accept(
                collector,
                ClassReader.SKIP_CODE or ClassReader.SKIP_DEBUG or ClassReader.SKIP_FRAMES
            )
            return collector.toMetadata()
        } finally {
            stream.close()
        }
    }

    /**
     * 这里只提取父类与接口签名，避免在 coremod 早期路径为层级判断构建完整的 ClassNode。
     */
    private class HierarchySignatureCollector : ClassVisitor(Opcodes.ASM5) {
        private var rawSuperClass: String? = null
        private var rawInterfaces: Array<out String> = emptyArray()

        override fun visit(
            version: Int,
            access: Int,
            name: String?,
            signature: String?,
            superName: String?,
            interfaces: Array<out String>
        ) {
            rawSuperClass = superName
            rawInterfaces = interfaces
        }

        fun toMetadata(): Metadata {
            val superClass = rawSuperClass?.let {
                toDotName(FMLDeobfuscatingRemapper.INSTANCE.map(it))
            }
            val interfaces = LinkedHashSet<String>(rawInterfaces.size)
            for (rawInterface in rawInterfaces) {
                interfaces += toDotName(FMLDeobfuscatingRemapper.INSTANCE.map(rawInterface))
            }
            return Metadata(superClass = superClass, interfaces = interfaces)
        }
    }
}
